### PR TITLE
Allow to use c++11 <cmath> before 10.7 by clang

### PR DESCRIPTION
**Pull request by catap**
See: https://trac.macports.org/ticket/65795

### CODE REVIEW

File: include/MacportsLegacySupport.h [hunk 1]
```diff
@@ -124,10 +124,11 @@
...
 #define __MP_LEGACY_SUPPORT_STAILQ_FOREACH__  (__APPLE__ && __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ < 1050)
 
-/* c++11 <cmath> PPC 10.[45] and Intel 10.[4-6], GNU g++ 4.6 through 8. */
-#if (__APPLE__ && __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ < 1070 \
-               && defined(__GNUC__) && (__GNUC__ <= 8)                 \
-               && ((__GNUC__ > 4) || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 6))))
+/* c++11 <cmath> PPC 10.[45] and Intel 10.[4-6], GNU g++ 4.6 through 8, or for any clang. */
+#if (__APPLE__ && __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ < 1070  \
+     && ( defined(__clang__)                                            \
+          || defined(__GNUC__) && (__GNUC__ <= 8)                       \
```
Comment (michaelld) [inline include/MacportsLegacySupport.h:130]: I would prefer to do `|| ( defined(__GNUC__) && ... && ... )` to make it clear that all of these clauses belong to the same compiler ...even though it's technically not necessary to do so

**Review by michaelld — APPROVED** [review]
This is fine with me ... prefer what I note but I can go either way. If @kencu likes this change then it's good to go IMHO

File: include/MacportsLegacySupport.h (hunk 1 above)
Comment (catap) [reply include/MacportsLegacySupport.h]: @michaelld I've tried to re-format condition to make it clear and easy to review.

Comment (catap) [reply include/MacportsLegacySupport.h]: To be honest I have no idea how to reformat it to make it easy to read :(

Comment (michaelld) [reply include/MacportsLegacySupport.h]: LOL I can't think of a good way either! This latest is good enough IMHO ... thanks!

**Comment by kencu** [comment]
I believe you would not want this pathway enabled when clang is building against ```-stdlib=libc++``` as that is already well tested and taken care of in the libc++ headers.

At least, I can't guess without testing how that would work out...

There is a simple preprocessor define in the libc++ headers somewhere to test for, if that is needed as an override.

**Comment by catap** [comment]
Ken, thanks, this is a good point.

**Comment by kencu** [comment]
now we just need some test proof that this works properly with all stdlib variations when using clang on 10.5.

Aside, we want this to work just like Apple’s math works on 10.7…. I have been meaning to check whether ll support is sensitive to setting c++11 with Apple’s compilers, and make sure we match them in this. We are more interested in matching Apple than a strict standard adherence on a 15 year old system ;)

**Comment by cjones051073** [comment]
Where do we stand on this now ? Are we happy to merge or is some more testing (by adding some test case here) required ?

**Comment by kencu** [comment]
I haven’t verified this by hand yet, but I suspect it will be right. I don’t think we can have an integrated test case because of the compiler requirement?

**Comment by kencu** [comment]
This appears to be still broken even with this PR, and there are a few other issues that turn up as well:

Here is the test.cpp file:
```
$ cat test.cpp
#include <cmath>

#if __MP_LEGACY_SUPPORT_CXX11_CMATH__
# warning "OVERRIDE is defined"
#endif

int main(void)
{
  long long int roundednumber;
  double number = 701.1;
  
  roundednumber = llround(number);

  return (0);
}
```

And here is the test.c file:
```
#include <math.h>

#if __MP_LEGACY_SUPPORT_CXX11_CMATH__
# warning "OVERRIDE is defined"
#endif


int main(void)
{
  long long int roundednumber;
  double number = 701.1;
  
  roundednumber = llround(number);

  return (0);
}
```



This is Apple's default compiler setup, which we don't want to change, but legacysupport does change it:

```
$ g++-4.2 test.cpp
$ g++-4.2 -std=c++98 test.cpp
test.cpp: In function ‘int main()’:
test.cpp:12: error: ‘llround’ was not declared in this scope
$ g++-4.2  -isystem/opt/local/include/LegacySupport  -std=c++98 test.cpp
```

similarly, clang++-7.0 should error in c++98 mode, but does not when legacysupport is used:
```
$ clang++-mp-7.0  test.cpp
$ clang++-mp-7.0  -std=c++98 test.cpp
test.cpp:12:19: error: use of undeclared identifier 'llround'; did you mean 'lround'?
  roundednumber = llround(number);
                  ^~~~~~~
                  lround
/usr/include/architecture/i386/math.h:388:17: note: 'lround' declared here
extern long int lround ( double );
                ^
1 error generated.
$ clang++-mp-7.0  -isystem/opt/local/include/LegacySupport -std=c++98  test.cpp
```


More importantly, clang++-mp-7.0 is still broken when building against macports-libstdc++, even if legacysupport is being used:

This version using -stdlib=libstdc++ works properly when legacysupport is added, as it should:

```
$  clang++-mp-7.0 -std=c++11  test.cpp
test.cpp:12:19: error: use of undeclared identifier 'llround'; did you mean 'lround'?
  roundednumber = llround(number);
                  ^~~~~~~
                  lround
/usr/include/architecture/i386/math.h:388:17: note: 'lround' declared here
extern long int lround ( double );
                ^
1 error generated.
$  clang++-mp-7.0  -isystem/opt/local/include/LegacySupport -std=c++11  test.cpp
```

This version using -stdlib=libc++ works properly with or without legacysupport, as it should:
```
$  clang++-mp-7.0 -stdlib=libc++ -std=c++11  test.cpp
$  clang++-mp-7.0 -isystem/opt/local/include/LegacySupport -stdlib=libc++ -std=c++11  test.cpp
```

But this version is broken, even if legacysupport is being used (which is one of the ones we were supposed to fix):
```
$  clang++-mp-7.0 -stdlib=macports-libstdc++ -std=c++11  test.cpp
In file included from test.cpp:1:
/opt/local/include/gcc/c++/cmath:1149:11: error: no member named 'llrint' in the global namespace; did you mean 'lrint'?
  using ::llrint;
        ~~^
/usr/include/architecture/i386/math.h:382:17: note: 'lrint' declared here
extern long int lrint ( double );
                ^
In file included from test.cpp:1:
/opt/local/include/gcc/c++/cmath:1150:11: error: no member named 'llrintf' in the global namespace; did you mean 'lrintf'?
  using ::llrintf;
        ~~^
/usr/include/architecture/i386/math.h:383:17: note: 'lrintf' declared here
extern long int lrintf ( float );
                ^
In file included from test.cpp:1:
/opt/local/include/gcc/c++/cmath:1151:11: error: no member named 'llrintl' in the global namespace; did you mean 'lrintl'?
  using ::llrintl;
        ~~^
/usr/include/architecture/i386/math.h:475:17: note: 'lrintl' declared here
extern long int lrintl(long double);
                ^
In file included from test.cpp:1:
/opt/local/include/gcc/c++/cmath:1153:11: error: no member named 'llround' in the global namespace; did you mean 'lround'?
  using ::llround;
        ~~^
/usr/include/architecture/i386/math.h:388:17: note: 'lround' declared here
extern long int lround ( double );
                ^
In file included from test.cpp:1:
/opt/local/include/gcc/c++/cmath:1154:11: error: no member named 'llroundf' in the global namespace; did you mean 'lroundf'?
  using ::llroundf;
        ~~^
/usr/include/architecture/i386/math.h:389:17: note: 'lroundf' declared here
extern long int lroundf ( float );
                ^
In file included from test.cpp:1:
/opt/local/include/gcc/c++/cmath:1155:11: error: no member named 'llroundl' in the global namespace; did you mean 'lroundl'?
  using ::llroundl;
        ~~^
/usr/include/architecture/i386/math.h:477:17: note: 'lroundl' declared here
extern long int lroundl(long double);
                ^
test.cpp:12:19: error: use of undeclared identifier 'llround'; did you mean 'lround'?
  roundednumber = llround(number);
                  ^~~~~~~
                  lround
/usr/include/architecture/i386/math.h:388:17: note: 'lround' declared here
extern long int lround ( double );
                ^
7 errors generated.
```
and this is not fixed by adding legacysupport:
```
$  clang++-mp-7.0 -isystem/opt/local/include/LegacySupport -stdlib=macports-libstdc++  -std=c++11  test.cppIn file included from test.cpp:1:In file included from /opt/local/include/LegacySupport/cmath:54:/opt/local/include/gcc/c++/cmath:1149:11: error: no member named 'llrint' in the global namespace; did you mean 'lrint'?
  using ::llrint;
        ~~^
/usr/include/architecture/i386/math.h:382:17: note: 'lrint' declared here
extern long int lrint ( double );
                ^
In file included from test.cpp:1:
In file included from /opt/local/include/LegacySupport/cmath:54:
/opt/local/include/gcc/c++/cmath:1150:11: error: no member named 'llrintf' in the global namespace; did you mean 'lrintf'?
  using ::llrintf;
        ~~^
/usr/include/architecture/i386/math.h:383:17: note: 'lrintf' declared here
extern long int lrintf ( float );
                ^
In file included from test.cpp:1:
In file included from /opt/local/include/LegacySupport/cmath:54:
/opt/local/include/gcc/c++/cmath:1151:11: error: no member named 'llrintl' in the global namespace; did you mean 'lrintl'?
  using ::llrintl;
        ~~^
/usr/include/architecture/i386/math.h:475:17: note: 'lrintl' declared here
extern long int lrintl(long double);
                ^
In file included from test.cpp:1:
In file included from /opt/local/include/LegacySupport/cmath:54:
/opt/local/include/gcc/c++/cmath:1153:11: error: no member named 'llround' in the global namespace; did you mean 'lround'?
  using ::llround;
        ~~^
/usr/include/architecture/i386/math.h:388:17: note: 'lround' declared here
extern long int lround ( double );
                ^
In file included from test.cpp:1:
In file included from /opt/local/include/LegacySupport/cmath:54:
/opt/local/include/gcc/c++/cmath:1154:11: error: no member named 'llroundf' in the global namespace; did you mean 'lroundf'?
  using ::llroundf;
        ~~^
/usr/include/architecture/i386/math.h:389:17: note: 'lroundf' declared here
extern long int lroundf ( float );
                ^
In file included from test.cpp:1:
In file included from /opt/local/include/LegacySupport/cmath:54:
/opt/local/include/gcc/c++/cmath:1155:11: error: no member named 'llroundl' in the global namespace; did you mean 'lroundl'?
  using ::llroundl;
        ~~^
/usr/include/architecture/i386/math.h:477:17: note: 'lroundl' declared here
extern long int lroundl(long double);
                ^
test.cpp:12:19: error: use of undeclared identifier 'llround'; did you mean 'lround'?
  roundednumber = llround(number);
                  ^~~~~~~
                  lround
/usr/include/architecture/i386/math.h:388:17: note: 'lround' declared here
extern long int lround ( double );
                ^
7 errors generated.
```

Also, using legacysupport appears to invoke the override when building with GCC7, both c and c++, even though in this case, ideally it shouldn't:
```
$ g++-mp-7 -std=c++11  test.cpp
$ g++-mp-7 -isystem/opt/local/include/LegacySupport  -std=c++11  test.cpp
test.cpp:4:3: warning: #warning "OVERRIDE is defined" [-Wcpp]
 # warning "OVERRIDE is defined"
   ^~~~~~~

```

```
$ gcc-mp-7 test.c
$ gcc-mp-7 -isystem/opt/local/include/LegacySupport test.c
test.c:4:3: warning: #warning "OVERRIDE is defined" [-Wcpp]
 # warning "OVERRIDE is defined"
   ^~~~~~~

**Comment by kencu** [comment]
do you have time to test it as above? I’m getting busy with other things, and we can’t commit it until it’s shown to work.

**Comment by catap** [comment]
@kencu I'll try to go over your comment and compare master and my changes in bahviour this week, probably very late to night.

**Comment by rmottola** [comment]
Hmm.. I just tried and I still have an issue building graphviz with clang-7, I see in MacportsLegacySupprt.h the mentioned ifdef's.

I wonder if __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ is set correctly? On 10.5 I sometimes had issues not having proper versions, but I think it is fine.

If use gcc7 instead of clang7, it works

**Comment by catap** [comment]
@rmottola I'm overload on another things and haven't got enough time to finish this one :(

**Comment by kencu** [comment]
hey, riccardo!

someone has to adjust the testing defines to make sure this define happens when building with clang against macports-libstdc++:

`
#define __MP_LEGACY_SUPPORT_CXX11_CMATH__ 1 `

and then verify it using the tests I outlined just above.

**Comment by kencu** [comment]
I am trying this out at present --  it is imperfect as we can't know which stdlib we are building against without including a std header first. Setting the stdlib in clang does not (AFAICT) define any intrinsic preprocessor macro that we could use.

```
/* c++11 <cmath> support. This was fixed properly on 10.7+, but generates errors on < 10.7 when using
   GNU g++ 4.6 through 6 or any clang - clang identifies itself as gcc 4.2.
   We can't test for the stdlib in use without having first included a stdlib header
*/
#if (__APPLE__ && __ENVIRONMENT_MAC_OS_X_VERSION_MIN_REQUIRED__ < 1070) \
               && (     ((__GNUC__ > 4) && (__GNUC__ <= 6))             \
                     || ((__GNUC__ == 4) && (__GNUC_MINOR__ >= 6))      \
                     || defined(__clang__)                              \
                  )
#define __MP_LEGACY_SUPPORT_CXX11_CMATH__ 1
#else
#define __MP_LEGACY_SUPPORT_CXX11_CMATH__ 0
#endif
```

**Comment by kencu** [comment]
It appears this PR is no longer needed due to this:

https://github.com/macports/macports-ports/commit/39844191e0d1a0c637b8af2c96058ae5c9c9c896

and I guess if we extend that to the gcc versions between 4.5 and 6, then the whole cmath fix in legacysupport can be removed.

**Pull request closed without merging**